ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/OnChainPrivacyPrecompiledContract.java

```diff
@@ -0,0 +1,444 @@
+
+  private final Enclave enclave;
+  private final WorldStateArchive privateWorldStateArchive;
+  private final PrivateStateStorage privateStateStorage;
+  private final PrivateStateRootResolver privateStateRootResolver;
+  private PrivateTransactionProcessor privateTransactionProcessor;
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  public OnChainPrivacyPrecompiledContract(
+      final GasCalculator gasCalculator, final PrivacyParameters privacyParameters) {
+    this(
+        gasCalculator,
+        privacyParameters.getEnclave(),
+        privacyParameters.getPrivateWorldStateArchive(),
+        privacyParameters.getPrivateStateStorage());
+  }
+
+  OnChainPrivacyPrecompiledContract(
+      final GasCalculator gasCalculator,
+      final Enclave enclave,
+      final WorldStateArchive worldStateArchive,
+      final PrivateStateStorage privateStateStorage) {
+    super("OnChainPrivacy", gasCalculator);
+    this.enclave = enclave;
+    this.privateWorldStateArchive = worldStateArchive;
+    this.privateStateStorage = privateStateStorage;
+    this.privateStateRootResolver = new PrivateStateRootResolver(privateStateStorage);
+  }
+
+  public void setPrivateTransactionProcessor(
+      final PrivateTransactionProcessor privateTransactionProcessor) {
+    this.privateTransactionProcessor = privateTransactionProcessor;
+  }
+
+  @Override
+  public Gas gasRequirement(final Bytes input) {
+    return Gas.of(0L);
+  }
+
+  @Override
+  public Bytes compute(final Bytes input, final MessageFrame messageFrame) {
+    final ProcessableBlockHeader currentBlockHeader = messageFrame.getBlockHeader();
+    if (!BlockHeader.class.isAssignableFrom(currentBlockHeader.getClass())) {
+      if (!messageFrame.isPersistingPrivateState()) {
+        // We get in here from block mining.
+        return Bytes.EMPTY;
+      } else {
+        throw new IllegalArgumentException(
+            "The MessageFrame contains an illegal block header type. Cannot persist private block metadata without current block hash.");
+      }
+    }
+    final Hash currentBlockHash = ((BlockHeader) currentBlockHeader).getHash();
+
+    final String key = input.slice(0, 32).toBase64String();
+
+    final ReceiveResponse receiveResponse;
+    try {
+      receiveResponse = enclave.receive(key);
+    } catch (final EnclaveClientException e) {
+      LOG.debug("Can not fetch private transaction payload with key {}", key, e);
+      return Bytes.EMPTY;
+    } catch (final EnclaveServerException e) {
+      LOG.error("Enclave is responding but errored perhaps it has a misconfiguration?", e);
+      throw e;
+    } catch (final EnclaveIOException e) {
+      LOG.error("Can not communicate with enclave is it up?", e);
+      throw e;
+    }
+
+    final BytesValueRLPInput bytesValueRLPInput =
+        new BytesValueRLPInput(
+            Bytes.wrap(Base64.getDecoder().decode(receiveResponse.getPayload())), false);
+    bytesValueRLPInput.enterList();
```

Comment:
This is here because we have changed the RLP to being a list that contains the private transaction and the version.
To read that here we should have a class (e.g. VersionedPrivateTransaction) that reads the list. The version should always be available for transactions that end up in this precompile!